runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/AbstractWindmillStream.java

```diff
@@ -49,46 +44,64 @@
  * stream if it is broken. Subclasses are responsible for retrying requests that have been lost on a
  * broken stream.
  *
- * <p>Subclasses should override onResponse to handle responses from the server, and onNewStream to
- * perform any work that must be done when a new stream is created, such as sending headers or
- * retrying requests.
+ * <p>Subclasses should override {@link #onResponse(ResponseT)} to handle responses from the server,
+ * and {@link #onNewStream()} to perform any work that must be done when a new stream is created,
+ * such as sending headers or retrying requests.
  *
- * <p>send and startStream should not be called from onResponse; use executor() instead.
+ * <p>{@link #send(RequestT)} and {@link #startStream()} should not be called from {@link
+ * #onResponse(ResponseT)}; use {@link #executeSafely(Runnable)} instead.
  *
  * <p>Synchronization on this is used to synchronize the gRpc stream state and internal data
  * structures. Since grpc channel operations may block, synchronization on this stream may also
  * block. This is generally not a problem since streams are used in a single-threaded manner.
  * However, some accessors used for status page and other debugging need to take care not to require
  * synchronizing on this.
+ *
+ * <p>{@link #start()} and {@link #shutdown()} are called once in the lifetime of the stream. Once
+ * {@link #shutdown()}, a stream in considered invalid and cannot be restarted/reused.
  */
 public abstract class AbstractWindmillStream<RequestT, ResponseT> implements WindmillStream {
 
-  public static final long DEFAULT_STREAM_RPC_DEADLINE_SECONDS = 300;
   // Default gRPC streams to 2MB chunks, which has shown to be a large enough chunk size to reduce
   // per-chunk overhead, and small enough that we can still perform granular flow-control.
   protected static final int RPC_STREAM_CHUNK_SIZE = 2 << 20;
-  private static final Logger LOG = LoggerFactory.getLogger(AbstractWindmillStream.class);
-  protected final AtomicBoolean clientClosed;
-  private final AtomicBoolean isShutdown;
-  private final AtomicLong lastSendTimeMs;
-  private final Executor executor;
+  // Indicates that the logical stream has been half-closed and is waiting for clean server
+  // shutdown.
+  private static final Status OK_STATUS = Status.fromCode(Status.Code.OK);
+  protected final Sleeper sleeper;
+
+  /**
+   * Used to guard {@link #start()} and {@link #shutdown()} behavior.
+   *
+   * @implNote Do not hold when performing IO. If also locking on {@code this} in the same context,
```

Comment:
if you acquire this before (this) lock, then this mutex could be blocked by I/O because we perform IO beneath this lock.

If this is supposed to be lightweight it seems like it should be acquired after (this) to avoid that.

Reply:
done